### PR TITLE
Funding pot contribution local

### PR DESCRIPTION
**Pull request by JeffreyJoel**
### Overview
This pull request is for the review of the `Funding_Pot_v1` contracts after adding implementation of `contrributeToTound()`

### Key Changes
- **Round Contribution:** Added logic to allow users contribute to rounds based on the specified access criteria`
- **Input Validation:** Added various custom errors to handle incorrect input validation and revert in case invalid input is passed
- **Unit Tests:**
  - Added unit and fuzz tests for `contributeToRound()` following the gherkin style method

### Focus Areas
- Provide feedback on input validation checks, any edge case leading to bug/security concern
-any changes with respect to Inverter coding standard

### CODE REVIEW

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 1]
```diff
@@ -182,6 +187,36 @@ contract LM_PC_FundingPot_v1 is
         }
     }
 
+    /// @inheritdoc ILM_PC_FundingPot_v1
+    function getRoundAccessCriteriaPrivilages(uint64 roundId_, uint8 accessId_)
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Not that important but privileges is spelt wrong

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 2]
```diff
@@ -182,6 +187,36 @@ contract LM_PC_FundingPot_v1 is
...
+            accessCriteriaPrivilages[roundId_][accessId_].overrideCap,
+            accessCriteriaPrivilages[roundId_][accessId_].start,
+            accessCriteriaPrivilages[roundId_][accessId_].cliff,
+            accessCriteriaPrivilages[roundId_][accessId_].end
+        );
+    }
+
     /// @inheritdoc ILM_PC_FundingPot_v1
     function getRoundCount() external view returns (uint64) {
         return nextRoundId;
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Is this accurate? If there's 1 round will this return 2 or 1?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 3]
```diff
@@ -237,4 +290,23 @@ interface ILM_PC_FundingPot_v1 is IERC20PaymentClientBase_v2 {
...
+    /// @param  overrideCap_ Whether to override the global cap
+    /// @param  _start Start timestamp for the access criteria
+    /// @param  _cliff Cliff timestamp for the access criteria
+    /// @param  _end End timestamp for the access criteria
+    function setAccessCriteriaPrivilages(
+        uint64 roundId_,
+        uint8 accessId_,
+        uint personalCap_,
+        bool overrideCap_,
+        uint _start,
```
Comment (leeftk) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: Underscore before the param? Is this correct? I may be wrong here so please double check

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 4]
```diff
@@ -937,6 +937,152 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+            roundId, accessId, personalCap, overrideCap, start, cliff, end
+        );
+        vm.stopPrank();
+    }
+
+    function testFuzzSetAccessCriteriaPrivilages_revertsGivenRoundDoesNotExist(
+        uint64 roundId,
+        uint8 accessId
+    ) public {
+        uint64 roundId = fundingPot.getRoundCount();
```
Comment (leeftk) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:905]: Does this actually pass? We have roundID in the params and then we're defining it and setting it to the roundcount, we're either defining it twice or the params are not actually fuzzing RoundID as it is being set to the value being returned from roundCount?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 5]
```diff
@@ -937,6 +937,152 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+                    .Module__LM_PC_FundingPot__RoundNotCreated
+                    .selector
+            )
+        );
+        fundingPot.setAccessCriteriaPrivilages(
+            roundId, accessId, personalCap, overrideCap, start, cliff, end
+        );
+    }
+
+    function testFuzzSetAccessCriteriaPrivilages_revertsGivenRoundIsActive()
```
Comment (leeftk) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: This isn't a fuzz test

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 6]
```diff
@@ -937,6 +937,152 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+                    .Module__LM_PC_FundingPot__RoundAlreadyStarted
+                    .selector
+            )
+        );
+        fundingPot.setAccessCriteriaPrivilages(
+            roundId, accessId, personalCap, overrideCap, start, cliff, end
+        );
+    }
+
+    function testFuzzSetAccessCriteriaPrivilages_revertsGivenAccessCriteriaIsOpen(
```
Comment (leeftk) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: This isn't a fuzz test :)

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 7]
```diff
@@ -937,6 +937,152 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+        );
+        (
+            bool isRoundOpen_,
+            uint personalCap_,
+            bool overrideCap_,
+            uint start_,
+            uint cliff_,
+            uint end_
+        ) = fundingPot.getRoundAccessCriteriaPrivilages(roundId, accessId);
+        assertEq(isRoundOpen_, accessCriteriaEnum == 0);
```
Comment (leeftk) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: Should the accessCriteriaEnum always equal zero?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 8]
```diff
@@ -1092,4 +1238,17 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
             }
         }
     }
+
+    function _helper_createAccessCriteriaPrivilages()
```
Comment (leeftk) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: I believe these values should be fuzzed but we should probs discuss this on a call

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 9]
```diff
@@ -374,6 +383,73 @@ contract LM_PC_FundingPot_v1 is
...
+        if (round.roundCap > 0 && totalRoundContribution >= round.roundCap) {
+            revert Module__LM_PC_FundingPot__RoundCapReached();
+        }
+        uint roundRemainingCap = round.roundCap - totalRoundContribution;
+        if (actualContributionAmount > roundRemainingCap) {
+            actualContributionAmount = roundRemainingCap;
+        }
+
+        // Transfer funds.
+        IERC20(contributionToken_).safeTransferFrom(
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Check and effects interaction should be followed here. ERC777, 1155 and many others perform a callback after transfer, since you're updating state after this call that could allow and attacker to drain all the funds :)

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 10]
```diff
@@ -374,6 +383,73 @@ contract LM_PC_FundingPot_v1 is
             _end
         );
     }
+
+    function contribute(
+        uint64 roundId_,
+        uint amount_,
+        uint8 accessId_,
+        address contributionToken_,
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Why would a user be able to specify the contribution token? I could just create a random ERC20 and then give myself a balance of 10000000 and contribute here? Is that right? I think this should be set during initialization

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 11]
```diff
@@ -374,6 +383,73 @@ contract LM_PC_FundingPot_v1 is
...
+
+        // Validate access criteria.
+        _validateAccessCriteria(roundId_, accessId_, merkleProof_);
+
+        // Retrieve user's previous contribution and calculate the personal cap.
+        address user = msg.sender;
+        uint userPreviousContribution = _getUserContribution(roundId_, user);
+        uint userPersonalCap = _getUserPersonalCap(roundId_, user);
+
+        if (userPreviousContribution >= userPersonalCap) {
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: This is incorrect, what we should be checking is that the users previous contribution PLUS the amount the currently want to deposit is less than the personal cap

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 12]
```diff
@@ -374,6 +383,73 @@ contract LM_PC_FundingPot_v1 is
...
+            revert Module__LM_PC_FundingPot__PersonalCapReached();
+        }
+
+        // Adjust contribution if it would exceed the personal cap.
+        uint userRemainingCap = userPersonalCap - userPreviousContribution;
+        uint actualContributionAmount = amount_;
+        if (amount_ > userRemainingCap) {
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: we're assuming that if the user can't deposit the amount the specified in the params that they'll want to deposit whatever they can and that shouldn't be the case.

If I go to a bank that limits my max deposit amount at 500 and I currently have 400 already in there. I got paid a check of 1000 and want to deposit the entire amount what this is going to do is deposit what I can (100) and then return 900 to me. What if I don't want to do that? What if I want to go to another bank and deposit the whole amount 1000?

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 13]
```diff
@@ -439,4 +515,123 @@ contract LM_PC_FundingPot_v1 is
...
+        bool accessGranted = false;
+        if (accessCriteria.accessCriteriaId == AccessCriteriaId.NFT) {
+            accessGranted =
+                _checkNftOwnership(accessCriteria.nftContract, msg.sender);
+        } else if (accessCriteria.accessCriteriaId == AccessCriteriaId.MERKLE) {
+            //TODO: Should I move this into a helper function
+
+            bytes32 leaf = keccak256(abi.encodePacked(msg.sender, roundId_));
+            accessGranted = MerkleProof.verify(
+                merkleProof_, accessCriteria.merkleRoot, leaf
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: can I use this merkle proof again? should we be storing this as used and not allow a user to use the same merkle proof twice??

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 14]
```diff
@@ -439,4 +515,123 @@ contract LM_PC_FundingPot_v1 is
...
+        if (accessCriteria.accessCriteriaId == AccessCriteriaId.OPEN) {
+            return;
+        }
+
+        bool accessGranted = false;
+        if (accessCriteria.accessCriteriaId == AccessCriteriaId.NFT) {
+            accessGranted =
+                _checkNftOwnership(accessCriteria.nftContract, msg.sender);
+        } else if (accessCriteria.accessCriteriaId == AccessCriteriaId.MERKLE) {
+            //TODO: Should I move this into a helper function
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: TODO left here

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 15]
```diff
@@ -439,4 +515,123 @@ contract LM_PC_FundingPot_v1 is
...
+    {
+        return userContributions[roundId_][user_];
+    }
+
+    function _getUserPersonalCap(uint64 roundId_, address user_)
+        internal
+        view
+        returns (uint)
+    {
+        uint basePersonalCap = 1000 ether;
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: why is the base personal cap 1000 ether? Where did this number come from

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 16]
```diff
@@ -439,4 +515,123 @@ contract LM_PC_FundingPot_v1 is
...
+    {
+        userContributions[roundId_][user_] += amount_;
+        roundTotalContributions[roundId_] += amount_;
+    }
+
+    function _checkAllowedAddressList(
+        address[] memory allowedAddresses,
+        address sender
+    ) internal pure returns (bool) {
+        for (uint i = 0; i < allowedAddresses.length; i++) {
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: if you store allowedAddressed.length before you run the for loop you'll save on gas as you wont have to read from state for each iteration of the for loop

**Review by leeftk — COMMENTED** [review]
Nice work @JeffreyJoel this is looking real good. There's still a lot of opportunity to learn and grow here but in time I think you'll pick these tricks up. We need to simplify the way we're writing code significantly and I'd say your validations could use a few reviews from yourself before sending over as you're adding a lot of redundancy. Keep up the hustle and let me know if you need to pair to go over any of this.

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 17]
```diff
@@ -571,27 +578,27 @@ contract LM_PC_FundingPot_v1 is
...
+            uint effectiveRoundCap = round_.roundCap;
 
             // If global accumulative caps are enabled, add unused capacity from previous rounds
-            if (round.globalAccumulativeCaps) {
+            if (round_.globalAccumulativeCaps) {
                 uint unusedCapacityFromPrevious = 0;
-                for (uint64 i = 1; i < roundId_; i++) {
+                for (uint64 i = 1; i < roundId_; ++i) {
                     Round storage prevRound = rounds[i];
                     if (!prevRound.globalAccumulativeCaps) continue;
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol:599]: I believe this is inverted, since we should be continuing if a round includes globalAccumlativeCaps, not if it doesn't.

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 18]
```diff
@@ -571,27 +578,27 @@ contract LM_PC_FundingPot_v1 is
...
     /// @param roundId_ ID of the round
     /// @param amount_ Requested contribution amount
-    /// @param round Round storage object
-    /// @param privileges Access criteria privileges
-    /// @param canOverrideTimeAndCap Whether the user can override cap constraints
+    /// @param round_ Round storage object
+    /// @param privileges_ Access criteria privileges
+    /// @param canOverrideTimeAndCap_ Whether the user can override cap constraints
     /// @return adjustedAmount The potentially adjusted contribution amount
     function _validateAndAdjustCaps(
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol:580]: We should be exposing this functions in the exposed contract and testing them internally. This is a great way to test these functions directly and make sure there's no bugs especially once we turn them to fuzz tests, my guess is with all this logics there's probably some edge cases in here.

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 19]
```diff
@@ -472,18 +478,18 @@ contract LM_PC_FundingPot_v1 is
...
     /// @dev    Validate uint start input.
-    /// @param  _start uint to validate.
-    /// @param  _cliff uint to validate.
-    /// @param  _end uint to validate.
+    /// @param  start_ uint to validate.
+    /// @param  cliff_ uint to validate.
+    /// @param  end_ uint to validate.
     /// @return True if uint is valid.
-    function _validTimes(uint _start, uint _cliff, uint _end)
+    function _validTimes(uint start_, uint cliff_, uint end_)
```
Comment (leeftk) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol:509]: Expose this in the exposed file and test, do this for all internal functions. Fuzz tests please :)

**Review by leeftk — COMMENTED** [review]
Added some feedback, mostly we need more tests for this and you probably want to rebase onto Jeffrey branch again as he added a bunch of changes.

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol (hunk 13 above)
Comment (leeftk) [reply src/modules/logicModule/LM_PC_FundingPot_v1.sol]: @JeffreyJoel please check this

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol (hunk 17 above)
Comment (JeffreyJoel) [reply src/modules/logicModule/LM_PC_FundingPot_v1.sol:599]: it skips the rounds where globalAccumlativeCaps is not enabled
